prov/shm/src/smr_msg.c

```diff
@@ -94,30 +94,97 @@ static ssize_t smr_generic_recv(struct smr_srx_ctx *srx, const struct iovec *iov
 		uint64_t tag, uint64_t ignore, uint64_t flags,
 		struct smr_queue *recv_queue, struct smr_queue *unexp_queue)
 {
-	struct smr_rx_entry *entry;
-	ssize_t ret = -FI_EAGAIN;
-	fi_addr_t use_addr;
+	struct smr_match_attr match_attr;
+	struct smr_rx_entry *rx_entry;
+	struct dlist_entry *dlist_entry;
+	int ret;
 
-	assert(iov_count <= SMR_IOV_LIMIT);
-	assert(!(flags & FI_MULTI_RECV) || iov_count == 1);
+	assert(iov_count <= SMR_IOV_LIMIT && !(flags & FI_MULTI_RECV));
+	addr = srx->dir_recv ? addr : FI_ADDR_UNSPEC;
 
-	use_addr = srx->dir_recv ? addr : FI_ADDR_UNSPEC;
-	ret = smr_check_unexp_queue(srx, iov, desc, iov_count, use_addr,
-				    context, tag, ignore, flags, unexp_queue);
-	if (!ret)
-		return ret;
+	match_attr.id = addr;
+	match_attr.ignore = ignore;
+	match_attr.tag = tag;
 
 	ofi_spin_lock(&srx->lock);
+	dlist_entry = dlist_remove_first_match(&unexp_queue->list,
+					       unexp_queue->match_func,
+					       &match_attr);
+	ofi_spin_unlock(&srx->lock);
+	if (!dlist_entry) {
+		rx_entry = smr_get_recv_entry(srx, iov, desc, iov_count, addr,
+					      context, tag, ignore, flags);
+		if (!rx_entry)
+			return -FI_ENOMEM;
+
+		dlist_insert_tail((struct dlist_entry *) (&rx_entry->peer_entry),
+				  &recv_queue->list);
+		return FI_SUCCESS;
+	}
+
+	rx_entry = container_of(dlist_entry, struct smr_rx_entry, peer_entry);
+	smr_init_rx_entry(rx_entry, iov, desc, iov_count, addr, context,
+			  tag, flags);
+
+	ret = srx->peer_srx.peer_ops->start_msg(&rx_entry->peer_entry);
+	if (ret)
+		return ret;
+
+	return FI_SUCCESS;
+}
 
-	entry = smr_get_recv_entry(srx, iov, desc, iov_count, use_addr, context,
-				   tag, ignore, flags);
-	if (!entry) {
+static ssize_t smr_generic_mrecv(struct smr_srx_ctx *srx, const struct iovec *iov,
+		void **desc, size_t iov_count, fi_addr_t addr, void *context,
+		uint64_t flags)
+{
+	struct smr_match_attr match_attr;
+	struct smr_rx_entry *rx_entry, *mrecv_entry;
+	struct dlist_entry *dlist_entry;
+	bool buf_done = false;
+	int ret;
+
+	assert(flags & FI_MULTI_RECV && iov_count == 1);
+	addr = srx->dir_recv ? addr : FI_ADDR_UNSPEC;
+
+	match_attr.id = addr;
+
+	ofi_spin_lock(&srx->lock);
+	mrecv_entry = smr_get_recv_entry(srx, iov, desc, iov_count, addr,
+					 context, 0, 0, flags);
+	mrecv_entry->peer_entry.size = ofi_total_iov_len(iov, iov_count);
+	if (!mrecv_entry) {
```

Comment:
It doesn't do any good to check that mrecv_entry is valid after the call before has already dereferenced it.